File: Coursera/Graphs/week1/p1.py


def explore(start, end, ar):
    global visited

    if start == end:
        return True

    for val in ar[start]:
        if visited[val]:
            continue
        visited[val] = True
        if explore(val, end, ar) == True:
            return True
    
    return False


n, m = map(int, input().split())
ar = []
visited = [False] * (n + 1)

for i in range(n + 1):
    ar.append([])

for i in range(m):
    x, y = map(int, input().split())
    ar[x].append(y)
    ar[y].append(x)

start, end = map(int, input().split())

res = explore(start, end, ar)
print("1" if res else "0")

File: Coursera/Graphs/week1/p1.cpp
#include <bits/stdc++.h>
using namespace std;

vector<bool> visited;

bool explore(int start, int end, vector<vector<int>> &arr)
{
    if (start == end)
        return true;
    for (int i : arr[start])
    {
        if (visited[i])
            continue;
        visited[i] = true;
        if (explore(i, end, arr))
            return true;
    }
    return false;
}

int main()
{
    int n, m, x, y;
    cin >> n >> m;

    visited.resize(n + 1, false);
    vector<vector<int>> points(n + 1, vector<int>());

    for (int i = 0; i < m; i++)
    {
        cin >> x >> y;
        points[x].emplace_back(y);
        points[y].emplace_back(x);
    }

    for (const auto &vec : points)
    {
        for (const auto value : vec)
            cout << value << ' ';
        cout << endl;
    }
}
